src/modules/tools/zprobe/CalibrationStrategy.cpp

```diff
@@ -0,0 +1,597 @@
+// Destroys A and b in the process
+// Result is stored in b
+template<typename T>
+void cholesky_backsub(int n, T A[/* n*n */], T b[/* n */]) {
+
+    // Compute LDL^T decomposition
+    for (int j = 0; j < n; j++) {
+        T inv_diagonal = 1;
+        for (int i = j; i < n; i++) {
+            T v = A[i*n + j];
+            for (int k = 0; k < j; k++) {
+                v -= A[k*n + j] * A[i*n + k];
+            }
+            if (i == j) {
+                A[i*n + j] = v;
+                inv_diagonal = 1/v;
+            } else {
+                A[j*n + i] = v; // store undivided value
+                A[i*n + j] = v * inv_diagonal;
+            }
+        }
+    }
+
+    // Compute A^-1*b, A=LDL^T
+    // L
+    for (int i = 0; i < n; i++) {
+        for (int j = 0; j < i; j++) {
+            b[i] -= A[i*n + j] * b[j];
+        }
+    }
+
+    // D
+    for (int i = 0; i < n; i++) {
+        b[i] /= A[i*n + i];
+    }
+
+    // L^T
+    for (int i = n-1; i >= 0; i--) {
+        for (int j = i+1; j < n; j++) {
+            b[i] -= A[j*n + i] * b[j];
+        }
+    }
+}
+
+bool CalibrationStrategy::probe_pattern(int n, int repeats, float actuator_positions[/*n*/][3]) {
+    if (n <= 7) {
+        return probe_symmetric(n, repeats, &actuator_positions[0]);
+    } else {
+        return probe_spiral(n, repeats, &actuator_positions[0]);
+    }
+}
+
+bool CalibrationStrategy::optimize_model(int n, int repeats, std::string const& parameters, StreamOutput* stream) {
+
+    std::vector<Vector3> actuator_positions(n);
+    std::vector<float> scratch;
+    std::vector<float> JTJ;
+    std::vector<float> JTr;
+    stream->printf("Probing %d points (%d repeat(s) per point).\n", n, repeats);
+
+    if (!probe_pattern(n, repeats, (float(*)[3])actuator_positions[0].data())) {
+        stream->printf("ERROR: Probing failed!\n");
+        return false;
+    }
+
+//    for (auto &v : actuator_positions) {
+//        stream->printf("%.5f %.5f %.5f\n",v[0],v[1],v[2]);
+//    }
+    float home_offs[3] = {};
+
+    {
+        void *returned_data;
+        if (PublicData::get_value(endstops_checksum, home_offset_checksum, &returned_data)) {
+            for (int i = 0; i < 3; i++) home_offs[i] = static_cast<float *>(returned_data)[i];
+        }
+    }
+
+    int m = parameters.size();
+
+    // Save initial values so we can roll back if we end up with a worse solution
+    std::vector<float> stored_values(m);
+    auto save = [&]() {
+        for (int i = 0; i < m; i++)
+            stored_values[i] = get_parameter(parameters[i]);
+    };
+    auto restore = [&]() {
+        for (int i = 0; i < m; i++)
+            set_parameter(parameters[i], stored_values[i]);
+    };
+    auto print_state = [&]() {
+        for (int i = 0; i < m; i++)
+            stream->printf("%c:%.3g ", parameters[i], stored_values[i]);
+        stream->printf("\n");
+    };
+
+    save();
+
+    float last_error = compute_model_rms_error(actuator_positions, home_offs[2]);
```

Comment:
home offset Z should be irrelevant here. It needs to be totally relative. My home offset is not relevant when the probe is attached, so that may be throwing stuff off. My Z home offset is set after probe is removed and after calibration is complete.